# Die Laufzeit der Binären Suche

<div class="prereq">
    <h3>Was man wissen sollte</h3>
    <div>
        Ohne die 
        <a class="prereq" href="/user-redirect/algoviz/lessons/04_Sortieren/01_BinaereSuche.ipynb">binäre Suche</a>
        zu kennen, braucht man hier gar nicht anfangen.
    </div>
</div>    

<div class="slideshow 04_Sortieren/02_BinaereSucheLaufzeit/slides.json">Die Laufzeit der binären Suche</div>

Im Folgenden wollen wir ein paar Experimente machen und die Zeit, dei die lineare und die binäre Suche benötigen messen. Dabei wollen wir insbesondere überprüfen, ob unsere Annahme, dass die benötigte Zeit ungefähr proportional zur Anzahl der gemachten Vergleiche bzw. Runden ist.

Als erste benötigen wir Operationen für das Füllen eines Arrays, die Ausgabe und natürlich die lineare und binäre Suche. Dazu nehmen wir die Implementierungen aus den beiden entsprechenden Lessons.

In [1]:
#include <iostream>
using namespace std;

In [2]:
void fuelleSortiert(int x[], int n) {
    srand(time(NULL));
    
    x[0] = rand() % 20;
    
    for ( int i = 1; i < n; i++ ) {
        x[i] =  x[i-1] + 1 + ( rand() % 20 );
    }
}

In [3]:
void ausgabe( int feld[], int laenge ) {
    
    for ( int pos = 0; pos < laenge; pos++ ) {
        cout << feld[pos] << " ";
    }
    cout << endl;
}

In die beiden Suchen bauen wir globale Zähler für die Vergleiche und Runden ein.

In [4]:
int vergleiche;
int runden;

In [5]:
int lineareSuche( int gesucht, int feld[], int laenge ) {
    
    vergleiche = 0;
    
    // Die Schleife
    for ( int pos = 0 ; pos < laenge; pos++ ) {
        vergleiche++;
        if ( feld[pos] == gesucht ) {
            // Breche die Schleife ab und gebe die Position zurück!
            return pos;
        }
    }
    
    return -1;
}

In [6]:
int binaereSuche(int gesucht, int daten[], int n) {
    int start = 0;
    int ende = n-1;
    int vergleich;
    
    vergleiche = 0;
    runden = 0;
    
    while ( start <= ende ) {
        runden++;
        
        vergleich = (start + ende)/2;
        vergleiche++;
        if ( daten[vergleich] == gesucht) {
            return vergleich;
        } else {
            vergleiche++;
            if ( daten[vergleich] < gesucht ) {
                start = vergleich+1;
            } else {
                ende = vergleich-1;
            }
        }
    }
    
    return -1;
}

Jetzt können wir ein Array sortiert befüllen und beide Operationen darauf suführen. Dabei wählen wir das gesuchte Element zufällig in dem Bereich zwischen 0 und `20*n` (zur Erinnerung: wir machen immer höchsten einen Schritt von 20 beim Befüllen).

In [35]:
srand(time(NULL));

int feld[20];
int gesucht = rand() % (20*20);

fuelleSortiert(feld,20);
ausgabe(feld,20);

cout << "gesucht: " << gesucht << endl << endl;

cout << "LINEARE SUCHE" << endl;
cout << "Index: " << lineareSuche(gesucht,feld,20) << endl;
cout << "Vergleiche: " << vergleiche << endl << endl;

cout << "BINAERE SUCHE" << endl;
cout << "Index: " << binaereSuche(gesucht,feld,20) << endl;
cout << "Vergleiche: " << vergleiche << endl;
cout << "Runden: " << runden << endl;

13 25 34 40 46 66 74 77 92 105 119 122 140 154 162 171 185 196 207 212 
gesucht: 153

LINEARE SUCHE
Index: -1
Vergleiche: 20

BINAERE SUCHE
Index: -1
Vergleiche: 10
Runden: 5


Mit dem ogen gegebenen Code können wir leicht die Anzahl der Vergleiche und Runden zwischen den beiden Algorithmen vergleichen. Dabei stellt man fest, dass tatsächlich die binäre Suche in der Regel deutlich weniger Runden und auch Vergleiche benötigt.

## Die Zeitmessung

Wir wollen aber auch die Zeit messen, die die Algorithmen benötigen und prüfen, ob sie proportional zu der Anzahl der Vergleiche bzw. Runden ist. Dazu füllen wir erst ein Array, das wir dann für die Experimente verwenden.

In [37]:
int feld[20];
int gesucht = rand() % (20*20);

fuelleSortiert(feld,20);
ausgabe(feld,20);

2 19 31 48 58 76 79 97 115 128 138 153 159 174 177 194 195 196 204 219 


In der globalen Variable `gesucht` legen wir den Wert fest, den wir suchen. Wir werden dabei mehrere Werte wählen, die im Array vorkommen und einen der nicht vorkommt.

In [96]:
int gesucht = 220;

Als nächstes Wollen wir wissen wie viele Vergleiche die lineare Suche benötigt.

In [97]:
lineareSuche(gesucht,feld,20);
cout << "Vergleiche (linear): " << vergleiche << endl << endl;

binaereSuche(gesucht,feld,20);
cout << "Vergleiche (binaer): " << vergleiche << endl << endl;

Vergleiche (linear): 20

Vergleiche (binaer): 10



Und jetzt kommt die Zeitmessung. Jupyter Notebook bietet die Möglichkeit die Zeit für die Ausführung einer Zelle zu messen. Dabei führt es die Zelle sogar mehrere Mal durch und ermittelt den Mittelwert der benötigten Zeit und die Standardabweichung. Der Befehl dafür ist `%%timeit`. Wird er am Anfang der Zelle ingefügt, wir die Messung durchgeführt. Das dauert durchaus etwas.

In [98]:
%%timeit
lineareSuche(gesucht,feld,20);

28.7 ns +- 0.0587 ns per loop (mean +- std. dev. of 7 runs 10000000 loops each)


In der Ausgabe sieht man die benötigte mittlere Zeit (in diesem Fall in der Regel in Nanosekunden), die Standardabweichung. Außerdem erfährt man, dass 7 Läufe mit jeweils 100000000 Durchführungen gemacht wurden (d.h. insgesamt 700000000 Suchen).

Und jetzt für die binäre Suche.

In [99]:
%%timeit
binaereSuche(gesucht,feld,20);

14.5 ns +- 0.0212 ns per loop (mean +- std. dev. of 7 runs 100000000 loops each)


<div class="task">
    <h3>Aufgabe</h3>
    <div>
        Diese Aufgabe ist etwas aufwändiger. Führen Sie die Zeitmessung für die Indices 0,5,9,10,14,15,17,18,19 und für ein Element, das nicht in <tt>feld[]</tt> enthalten ist. D.h. sie setzen <tt>gesucht</tt> auf den Wert am entsprechenden Index. Notieren Sie sich die Anzahl der benötigten Vergleiche und die mittlere benötigte Zeit. Notieren Sie beides getrennt für die lineare und binäre Suche.
    </div>
    <div>
        Berechnen Sie für die Messpaare die Zeit pro Vergleich und vergleichen Sie die erhaltenen Werte.
    </div>
</div>    

Hier sind die Werte, die ich bei einer Testmessung für die **lineare Suche** erhalten habe.

<table>
<tr>
 <th>Vergleiche</th><td>1</td><td>6</td><td>10</td><td>11</td><td>15</td><td>18</td><td>19</td><td>20</td>
</tr>
<tr>
 <th>Zeit [ns]</th><td>2.35</td><td>8.12</td><td>12.9</td><td>14.4</td><td>20.4</td><td>25.2</td><td>26.6</td><td>28.7</td>
</tr>
<tr>
 <th>Zeit pro Vergleich[ns]</th>
    <td>2.35</td>
    <td>1.35</td>
    <td>1.29</td>
    <td>1.3</td>
    <td>1.36</td>
    <td>1.4</td>
    <td>1.4</td>
    <td>1.44</td>
</tr>
</table>

Wie man sieht ist die Zeit pro Vergleich in etwa konstant, was für eine Proportionalität zwischen den beiden Größen Laufzeit und Anzahl der Vergleiche spricht. Allerings ist das natürlich noch kein echter Nachweis. Dazu müsste man einen viel größere Experimentalreihe durchführen und die Messwerte genauer auswerten.

Und hier das Ganze für die **binäre Suche**.

<table>
<tr>
 <th>Vergleiche</th>
    <td>1</td>
    <td>3</td>
    <td>5</td>
    <td>7</td>
    <td>7</td>
    <td>7</td>
    <td>7</td>
    <td>10</td>
</tr>
<tr>
 <th>Zeit [ns]</th>
    <td>3.59</td>
    <td>5.56</td>
    <td>8.07</td>
    <td>11.3</td>
    <td>11.3</td>
    <td>10.6</td>
    <td>10.8</td>
    <td>14.5</td>
</tr>
<tr>
 <th>Zeit pro Vergleich[ns]</th>
    <td>3.59</td>
    <td>1.86</td>
    <td>1.61</td>
    <td>1.61</td>
    <td>1.61</td>
    <td>1.51</td>
    <td>1.54</td>
    <td>1.45</td>
</tr>
</table>

Auch hier scheint sich die Proportionalität im Wesentlichen zu bestätigen. Was aber auffällt ist, dass die Zeit pro Vergleich etwa höher zu ein scheint. Dies könnte man z.B. damit erklären, das die Anweisungen, die neben den Vergleichen ausgeführt werden, bei der binären Suche etwas länger benötigen, als die bei der linearen Suche. Trotzdem scheint die geringere Anzahl der Vergleiche bei der binären Suche dazu zu führen, dass sie im Algemeinen schneller ist.

Insgesamt scheinen unsere sehr einfchen Experimente die Vermutung zu bestätigen, dass die Antzahl der Vergleiche die Laufzeit **dominiert**, oder anders ausgedrückt: Die Laufzeit ist im Wesentlichen proportional zur Anzahl der Vergleiche.

<div class="followup">
    <h3>Wo es weiter geht</h3>
    <div>
        Im Folgenden werden wir dieses Herangehen mathematisch präzisieren. Dazu werden wir schrittweise das
        Konzept der <a class="followup" href="/user-redirect/algoviz/lessons/04_Sortieren/03_AsymptotischeLaufzeiten.ipynb">Asymptotischen Laufzeit</a> einführen.
    </div>
</div>    